src/target/riscv/riscv.c

```diff
@@ -462,11 +462,109 @@ static void trigger_from_breakpoint(struct trigger *trigger,
 	trigger->unique_id = breakpoint->unique_id;
 }
 
-static int maybe_add_trigger_t1(struct target *target,
-		struct trigger *trigger, uint64_t tdata1)
+static bool match_napot_checker(struct trigger *trigger, riscv_reg_t *tdata2)
+{
+	riscv_reg_t addr = trigger->address;
+	riscv_reg_t size = trigger->length;
+	if (size > 1 &&
+		(size & (size - 1)) == 0 &&
+		(addr & (size - 1)) == 0) {
+		if (tdata2)
+			*tdata2 = addr | ((size - 1) >> 1);
+		return true;
+	}
+	return false;
+}
+
+static int find_trigger(struct target *target, int type, int num, int *idx)
 {
 	RISCV_INFO(r);
 
+	riscv_reg_t tselect;
+	if (riscv_get_register(target, &tselect, GDB_REGNO_TSELECT) != ERROR_OK)
+		return ERROR_FAIL;
+
+	unsigned int i;
+	int found = 0, cnt = 0;
+	for (i = 0; i < r->trigger_count; i++) {
+		if (r->trigger_unique_id[i] == -1) {
+			riscv_reg_t tdata1 = 0;
+			riscv_set_register(target, GDB_REGNO_TSELECT, i);
+			if (riscv_get_register(target, &tdata1, GDB_REGNO_TDATA1) == ERROR_OK) {
+				int t = get_field(tdata1, CSR_TDATA1_TYPE(riscv_xlen(target)));
+				if (type == t) {
+					found = (++cnt == num);
+					if (found)
+						break;
+					continue;
+				}
+			}
+		}
+		cnt = 0;
+	}
+
+	riscv_set_register(target, GDB_REGNO_TSELECT, tselect);
+
+	if (found) {
+		if (idx)
+			*idx = i - (num - 1);
+		return ERROR_OK;
+	}
+
+	return ERROR_TARGET_RESOURCE_NOT_AVAILABLE;
+}
+
+static int get_trigger(struct target *target, int idx, riscv_reg_t *tdata1, riscv_reg_t *tdata2)
```

Comment:
This function, set_trigger, and find_trigger, all save/restore tselect. That's a lot of wasted time. Instead, can you comment that these functions clobber tselect, and rely on add_trigger()/remove_trigger() to do the save/restore?